src/main/java/frc/robot/RobotContainer.java

```diff
@@ -1,26 +1,77 @@
+import edu.wpi.first.math.geometry.Transform3d;
 import edu.wpi.first.wpilibj2.command.Command;
-import frc.robot.subsystems.example.*;
+import edu.wpi.first.wpilibj2.command.button.CommandXboxController;
+import frc.robot.subsystems.example.ExampleSubsystem;
+import frc.robot.subsystems.example.ExampleSubsystemIO;
+import frc.robot.subsystems.example.ExampleSubsystemIOReal;
+import frc.robot.subsystems.example.ExampleSubsystemIOSim;
+import frc.robot.swerve.*;
+import frc.robot.vision.PhotonVisionIOReal;
+import frc.robot.vision.Vision;
+import frc.robot.vision.VisionModule;
+import frc.robot.vision.VisionSimIO;
+import org.photonvision.PhotonCamera;
+import org.photonvision.simulation.SimCameraProperties;
 
 public class RobotContainer {
 
     private static RobotContainer INSTANCE = null;
 
+    private final SwerveDrive swerveDrive;
+    private final CommandXboxController xboxController = new CommandXboxController(0);
+
     /** The container for the robot. Contains subsystems, OI devices, and commands. */
     private RobotContainer() {
         ExampleSubsystemIO exampleSubsystemIO;
+        ModuleIO[] moduleIOs = new ModuleIO[4];
+        GyroIO gyroIO;
+        VisionModule frontLeftVisionModule;
         switch (Constants.CURRENT_MODE) {
             case REAL:
                 exampleSubsystemIO = new ExampleSubsystemIOReal();
+                for (int i = 0; i < moduleIOs.length; i++) {
+                    moduleIOs[i] =
+                            new ModuleIOTalonFX(
+                                    Ports.SwerveDrive.DRIVE_IDS[i],
+                                    Ports.SwerveDrive.ANGLE_IDS[i],
+                                    Ports.SwerveDrive.ENCODER_IDS[i],
+                                    SwerveConstants.DRIVE_MOTOR_CONFIGS,
+                                    SwerveConstants.ANGLE_MOTOR_CONFIGS);
+                }
+                gyroIO = new GyroIOReal();
+                frontLeftVisionModule =
+                        new VisionModule(
+                                new PhotonVisionIOReal(
+                                        new PhotonCamera("Front left camera"),
+                                        new Transform3d(0.293, 0.293, 0.2, new Rotation3d()),
+                                        AprilTagFields.k2024Crescendo.loadAprilTagLayoutField()));
                 break;
             case SIM:
             case REPLAY:
             default:
                 exampleSubsystemIO = new ExampleSubsystemIOSim();
+                for (int i = 0; i < moduleIOs.length; i++) {
+                    moduleIOs[i] = new ModuleIOSim();
+                }
+                gyroIO = new GyroIOSim();
+                frontLeftVisionModule =
+                        new VisionModule(
+                                new VisionSimIO(
+                                        new PhotonCamera("Front left camera"),
+                                        new Transform3d(0.293, 0.293, 0.2, new Rotation3d()),
```

Comment:
https://github.com/Galaxia5987/robot-2024/blob/49e1d00b7ee4bd137827cabb989bd0b2945a987f/src/main/java/frc/robot/RobotContainer.java#L64-L65
Duplicate, also should be in a different constant file (why not do that in another file)?

Reply:
4b493526b5b891e1af2dd6cf7c25df441ef73c52